core/commands/add.go

```diff
@@ -61,47 +94,92 @@ remains to be implemented.
 					return
 				}
 
-				_, err = addFile(n, file, outChan)
+				_, err = addFile(n, file, outChan, progress)
 				if err != nil {
 					return
 				}
 			}
 		}()
-
-		return outChan, nil
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			outChan, ok := res.Output().(<-chan interface{})
-			if !ok {
-				return nil, u.ErrCast()
-			}
+	PostRun: func(res cmds.Response) {
+		outChan, ok := res.Output().(<-chan interface{})
+		if !ok {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+		res.SetOutput(nil)
 
-			quiet, _, err := res.Request().Option("quiet").Bool()
-			if err != nil {
-				return nil, err
+		quiet, _, err := res.Request().Option("quiet").Bool()
+		if err != nil {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+
+		size := int64(0)
+		s, found := res.Request().Values()["size"]
+		if found {
+			size = s.(int64)
+		}
+		showProgressBar := size >= progressBarMinSize
+
+		var bar *pb.ProgressBar
+		var terminalWidth int
+		if showProgressBar {
+			bar = pb.New64(size).SetUnits(pb.U_BYTES)
+			bar.ManualUpdate = true
+			bar.Start()
+
+			// the progress bar lib doesn't give us a way to get the width of the output,
+			// so as a hack we just use a callback to measure the output, then git rid of it
+			terminalWidth = 0
+			bar.Callback = func(line string) {
+				terminalWidth = len(line)
+				bar.Callback = nil
+				bar.Output = os.Stderr
+				log.Infof("terminal width: %v\n", terminalWidth)
 			}
+			bar.Update()
+		}
```

Comment:
Generalizing the progress bar stuff is a lot harder that it looks, since the commands differ a lot. And the parts that are generalizable are the parts that are already easy to call, so I don't think it's really worth it right now (just adds complexity to the commands lib).